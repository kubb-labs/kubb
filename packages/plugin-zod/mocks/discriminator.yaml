---
openapi: 3.0.1
info:
  title: Example Schema
servers:
- url: http://localhost/
paths:
  "/example":
    get:
      responses:
        '200':
          description: Ok
          content:
            application/json:
              schema:
                "$ref": "#/components/schemas/Advanced"
components:
  schemas:
    Advanced:
      title: ValueSpecification
      required:
        - type
      type: object
      oneOf:
        - $ref: "#/components/schemas/EnumerationValueSpecificationDto"
        - $ref: "#/components/schemas/RangeValueSpecificationDto"
        - $ref: "#/components/schemas/RegexValueSpecificationDto"
        - $ref: "#/components/schemas/SliderValueSpecificationDto"
      properties:
        type:
          minLength: 1
          enum:
            - enum
            - range
            - regex
            - slider
          type: string
          readOnly: true
      additionalProperties: false
      discriminator:
        propertyName: type
        mapping:
          enum: "#/components/schemas/EnumerationValueSpecificationDto"
          range: "#/components/schemas/RangeValueSpecificationDto"
          regex: "#/components/schemas/RegexValueSpecificationDto"
          slider: "#/components/schemas/SliderValueSpecificationDto"
    EnumerationValueSpecificationDto:
      title: EnumerationValueSpecification
      required:
        - enumeration
        - type
      type: object
      properties:
        type:
          minLength: 1
          type: string
          readOnly: true
        enumeration:
          type: array
          items: { }
      additionalProperties: false

    RangeValueSpecificationDto:
      title: RangeValueSpecification
      required:
        - max
        - min
        - type
      type: object
      properties:
        type:
          minLength: 1
          type: string
          readOnly: true
        min:
          type: number
          format: float
        max:
          type: number
          format: float
      additionalProperties: false

    RegexValueSpecificationDto:
      title: RegexValueSpecification
      required:
        - regex
        - type
      type: object
      properties:
        type:
          minLength: 1
          type: string
          readOnly: true
        regex:
          minLength: 1
          type: string
      additionalProperties: false


    SliderValueSpecificationDto:
      title: SliderValueSpecification
      required:
        - left_text
        - max_value
        - right_text
        - type
      type: object
      properties:
        type:
          minLength: 1
          type: string
          readOnly: true
        left_text:
          $ref: "#/components/schemas/LocalizedTextDto"
        right_text:
          $ref: "#/components/schemas/LocalizedTextDto"
        max_value:
          type: integer
          format: int32
      additionalProperties: false

    LocalizedTextDto:
      title: LocalizedText
      required:
        - default_text
        - text_id
      type: object
      properties:
        text_id:
          minLength: 1
          type: string
        default_text:
          minLength: 1
          type: string
      additionalProperties: false
