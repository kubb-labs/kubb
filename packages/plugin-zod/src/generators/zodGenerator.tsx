import { type OperationSchema as OperationSchemaType, SchemaGenerator, createReactGenerator } from '@kubb/plugin-oas'
import { Operations, OperationSchema, Schema } from '../components'
import type { PluginZod } from '../types'
import {File, useApp} from "@kubb/react";
import {useSchemaManager} from "@kubb/plugin-oas/hooks";
import {pluginZodName} from "../plugin.ts";
import {pluginTsName} from "@kubb/plugin-ts";

export const zodGenerator = createReactGenerator<PluginZod>({
  name: 'plugin-zod',
  Operations({ options }) {
    if (!options.operations) {
      return null
    }

    return <Operations.File />
  },
  Operation() {
    return <OperationSchema.File />
  },
  Schema({ schema, name, tree }) {
    const {
      pluginManager,
      plugin: {
        options: {infer, typed,  mapper, importPath },
      },
      mode,
    } = useApp<PluginZod>()
    const { getFile, getImports } = useSchemaManager()

    const file = getFile(name)
    const imports = getImports(tree)

    const resolvedName = pluginManager.resolveName({
      name,
      pluginKey: [pluginZodName],
      type: 'function',
    })

    // used for this.options.typed
    const typedName = pluginManager.resolveName({
      name,
      pluginKey: [pluginTsName],
      type: 'type',
    })

    const typedBaseName = pluginManager.resolveName({
      name: name,
      pluginKey: [pluginTsName],
      type: 'file',
    })

    const typedPath = pluginManager.resolvePath({
      baseName: typedBaseName,
      pluginKey: [pluginTsName],
    })


    const inferTypedName = pluginManager.resolveName({
      name,
      pluginKey: [pluginZodName],
      type: 'type',
    })

    return (
      <File baseName={file.baseName} path={file.path} meta={file.meta}>
        <File.Import name={['z']} path={importPath} />
        {typed && typedName && typedPath && <File.Import isTypeOnly root={file.path} path={typedPath} name={[typedName]} />}

        {mode === 'split' && imports.map((imp, index) => <File.Import key={index} root={file.path} path={imp.path} name={imp.name} />)}
        <Schema
          name={resolvedName}
          typedName={typed? typedName: undefined}
          inferTypedName={infer? inferTypedName: undefined}
          description={schema.description}
          tree={tree}
          mapper={mapper}
        />
      </File>
    )
  },
})
