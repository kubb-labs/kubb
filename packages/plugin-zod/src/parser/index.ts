import transformers from '@kubb/core/transformers'
import { type SchemaKeywordMapper, type SchemaTree, isKeyword, schemaKeywords } from '@kubb/plugin-oas'

import type { Schema, SchemaKeywordBase, SchemaMapper } from '@kubb/plugin-oas'

const zodKeywordMapper = {
  any: () => 'z.any()',
  unknown: () => 'z.unknown()',
  number: (coercion?: boolean, min?: number, max?: number) => {
    return [coercion ? 'z.coerce.number()' : 'z.number()', min !== undefined ? `.min(${min})` : undefined, max !== undefined ? `.max(${max})` : undefined]
      .filter(Boolean)
      .join('')
  },
  integer: (coercion?: boolean, min?: number, max?: number) => {
    return [
      coercion ? 'z.coerce.number().int()' : 'z.number().int()',
      min !== undefined ? `.min(${min})` : undefined,
      max !== undefined ? `.max(${max})` : undefined,
    ]
      .filter(Boolean)
      .join('')
  },
  object: (value?: string) => `z.object({${value}})`,
  string: (coercion?: boolean, min?: number, max?: number) => {
    return [coercion ? 'z.coerce.string()' : 'z.string()', min !== undefined ? `.min(${min})` : undefined, max !== undefined ? `.max(${max})` : undefined]
      .filter(Boolean)
      .join('')
  },
  boolean: () => 'z.boolean()',
  undefined: () => 'z.undefined()',
  nullable: () => '.nullable()',
  null: () => 'z.null()',
  nullish: () => '.nullish()',
  array: (items: string[] = [], min?: number, max?: number) => {
    return [`z.array(${items?.join('')})`, min !== undefined ? `.min(${min})` : undefined, max !== undefined ? `.max(${max})` : undefined]
      .filter(Boolean)
      .join('')
  },
  tuple: (items: string[] = []) => `z.tuple([${items?.join(', ')}])`,
  enum: (items: string[] = []) => `z.enum([${items?.join(', ')}])`,
  union: (items: string[] = []) => `z.union([${items?.join(', ')}])`,
  const: (value?: string | number | boolean) => `z.literal(${value ?? ''})`,
  /**
   * ISO 8601
   */
  datetime: (offset = false, local = false) => {
    if (offset) {
      return `z.string().datetime({ offset: ${offset} })`
    }

    if (local) {
      return `z.string().datetime({ local: ${local} })`
    }

    return 'z.string().datetime()'
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO date format (YYYY-MM-DD)
   * @default ISO date format (YYYY-MM-DD)
   */
  date: (type: 'date' | 'string' = 'string', coercion?: boolean) => {
    if (type === 'string') {
      return 'z.string().date()'
    }

    if (coercion) {
      return 'z.coerce.date()'
    }

    return 'z.date()'
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO time format (HH:mm:ss[.SSSSSS])
   * @default ISO time format (HH:mm:ss[.SSSSSS])
   */
  time: (type: 'date' | 'string' = 'string', coercion?: boolean) => {
    if (type === 'string') {
      return 'z.string().time()'
    }

    if (coercion) {
      return 'z.coerce.date()'
    }

    return 'z.date()'
  },
  uuid: (coercion?: boolean) => (coercion ? 'z.coerce.string().uuid()' : 'z.string().uuid()'),
  url: (coercion?: boolean) => (coercion ? 'z.coerce.string().url()' : 'z.string().url()'),
  strict: () => '.strict()',
  default: (value?: string | number | true) => `.default(${value ?? ''})`,
  and: (items: string[] = []) => items?.map((item) => `.and(${item})`).join(''),
  describe: (value = '') => `.describe(${value})`,
  min: (value?: number) => `.min(${value ?? ''})`,
  max: (value?: number) => `.max(${value ?? ''})`,
  optional: () => '.optional()',
  matches: (value = '', coercion?: boolean) => (coercion ? `z.coerce.string().regex(${value})` : `z.string().regex(${value})`),
  email: (coercion?: boolean) => (coercion ? 'z.coerce.string().email()' : 'z.string().email()'),
  firstName: undefined,
  lastName: undefined,
  password: undefined,
  phone: undefined,
  readOnly: undefined,
  writeOnly: undefined,
  ref: (value?: string) => (value ? `z.lazy(() => ${value})` : undefined),
  blob: () => 'z.instanceof(File)',
  deprecated: undefined,
  example: undefined,
  schema: undefined,
  catchall: (value?: string) => (value ? `.catchall(${value})` : undefined),
  name: undefined,
} satisfies SchemaMapper<string | null | undefined>

/**
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

export function sort(items?: Schema[]): Schema[] {
  const order: string[] = [
    schemaKeywords.string,
    schemaKeywords.datetime,
    schemaKeywords.date,
    schemaKeywords.time,
    schemaKeywords.tuple,
    schemaKeywords.number,
    schemaKeywords.object,
    schemaKeywords.enum,
    schemaKeywords.url,
    schemaKeywords.email,
    schemaKeywords.firstName,
    schemaKeywords.lastName,
    schemaKeywords.password,
    schemaKeywords.matches,
    schemaKeywords.uuid,
    schemaKeywords.min,
    schemaKeywords.max,
    schemaKeywords.default,
    schemaKeywords.describe,
    schemaKeywords.optional,
    schemaKeywords.nullable,
    schemaKeywords.nullish,
    schemaKeywords.null,
  ]

  if (!items) {
    return []
  }

  return transformers.orderBy(items, [(v) => order.indexOf(v.keyword)], ['asc'])
}

const shouldCoerce = (coercion: ParserOptions['coercion'] | undefined, type: 'dates' | 'strings' | 'numbers'): boolean => {
  if (coercion === undefined) {
    return false
  }
  if (typeof coercion === 'boolean') {
    return coercion
  }

  return !!coercion[type]
}

type ParserOptions = {
  name: string
  typeName?: string
  description?: string
  keysToOmit?: string[]
  mapper?: Record<string, string>
  coercion?: boolean | { dates?: boolean; strings?: boolean; numbers?: boolean }
}

export function parse({ parent, current, siblings }: SchemaTree, options: ParserOptions): string | undefined {
  const value = zodKeywordMapper[current.keyword as keyof typeof zodKeywordMapper]

  if (!value) {
    return undefined
  }

  if (isKeyword(current, schemaKeywords.union)) {
    // zod union type needs at least 2 items
    if (Array.isArray(current.args) && current.args.length === 1) {
      return parse({ parent, current: current.args[0] as Schema, siblings }, options)
    }
    if (Array.isArray(current.args) && !current.args.length) {
      return ''
    }

    return zodKeywordMapper.union(
      sort(current.args)
        .map((schema, _index, siblings) => parse({ parent: current, current: schema, siblings }, options))
        .filter(Boolean),
    )
  }

  if (isKeyword(current, schemaKeywords.and)) {
    const items = sort(current.args)
      .filter((schema: Schema) => {
        return ![schemaKeywords.optional, schemaKeywords.describe].includes(schema.keyword as typeof schemaKeywords.describe)
      })
      .map((schema: Schema, _index, siblings) => parse({ parent: current, current: schema, siblings }, options))
      .filter(Boolean)

    return `${items.slice(0, 1)}${zodKeywordMapper.and(items.slice(1))}`
  }

  if (isKeyword(current, schemaKeywords.array)) {
    return zodKeywordMapper.array(
      sort(current.args.items)
        .map((schemas, _index, siblings) => parse({ parent: current, current: schemas, siblings }, options))
        .filter(Boolean),
      current.args.min,
      current.args.max,
    )
  }

  if (isKeyword(current, schemaKeywords.enum)) {
    if (current.args.asConst) {
      if (current.args.items.length === 1) {
        const child = {
          keyword: schemaKeywords.const,
          args: current.args.items[0],
        }
        return parse({ parent: current, current: child, siblings: [child] }, options)
      }

      return zodKeywordMapper.union(
        current.args.items
          .map((schema) => ({
            keyword: schemaKeywords.const,
            args: schema,
          }))
          .map((schema, _index, siblings) => {
            return parse({ parent: current, current: schema, siblings }, options)
          })
          .filter(Boolean),
      )
    }

    return zodKeywordMapper.enum(
      current.args.items.map((schema) => {
        if (schema.format === 'boolean') {
          return transformers.stringify(schema.value)
        }

        if (schema.format === 'number') {
          return transformers.stringify(schema.value)
        }
        return transformers.stringify(schema.value)
      }),
    )
  }

  if (isKeyword(current, schemaKeywords.ref)) {
    return zodKeywordMapper.ref(current.args?.name)
  }

  if (isKeyword(current, schemaKeywords.object)) {
    const properties = Object.entries(current.args?.properties || {})
      .filter((item) => {
        const schema = item[1]
        return schema && typeof schema.map === 'function'
      })
      .map(([name, schemas]) => {
        const nameSchema = schemas.find((schema) => schema.keyword === schemaKeywords.name) as SchemaKeywordMapper['name']
        const mappedName = nameSchema?.args || name

        // custom mapper(pluginOptions)
        if (options.mapper?.[mappedName]) {
          return `"${name}": ${options.mapper?.[mappedName]}`
        }

        return `"${name}": ${sort(schemas)
          .map((schema, array, siblings) => parse({ parent: current, current: schema, siblings }, options))
          .filter(Boolean)
          .join('')}`
      })
      .join(',')

    const additionalProperties = current.args?.additionalProperties?.length
      ? current.args.additionalProperties
          .map((schema, _index, siblings) => parse({ parent: current, current: schema, siblings }, options))
          .filter(Boolean)
          .join('')
      : undefined

    const text = [
      zodKeywordMapper.object(properties),
      current.args?.strict ? zodKeywordMapper.strict() : undefined,
      additionalProperties ? zodKeywordMapper.catchall(additionalProperties) : undefined,
    ].filter(Boolean)

    return text.join('')
  }

  if (isKeyword(current, schemaKeywords.tuple)) {
    return zodKeywordMapper.tuple(
      current.args.items.map((schema, _index, siblings) => parse({ parent: current, current: schema, siblings }, options)).filter(Boolean),
    )
  }

  if (isKeyword(current, schemaKeywords.const)) {
    if (current.args.format === 'number' && current.args.value !== undefined) {
      return zodKeywordMapper.const(Number.parseInt(current.args.value?.toString()))
    }

    if (current.args.format === 'boolean' && current.args.value !== undefined) {
      return zodKeywordMapper.const(current.args.value)
    }
    return zodKeywordMapper.const(transformers.stringify(current.args.value))
  }

  if (isKeyword(current, schemaKeywords.matches)) {
    if (current.args) {
      return zodKeywordMapper.matches(transformers.toRegExpString(current.args), shouldCoerce(options.coercion, 'strings'))
    }
  }

  if (isKeyword(current, schemaKeywords.default)) {
    if (current.args) {
      return zodKeywordMapper.default(current.args)
    }
  }

  if (isKeyword(current, schemaKeywords.describe)) {
    if (current.args) {
      return zodKeywordMapper.describe(transformers.stringify(current.args.toString()))
    }
  }

  if (isKeyword(current, schemaKeywords.string)) {
    return zodKeywordMapper.string(shouldCoerce(options.coercion, 'strings'))
  }

  if (isKeyword(current, schemaKeywords.uuid)) {
    return zodKeywordMapper.uuid(shouldCoerce(options.coercion, 'strings'))
  }

  if (isKeyword(current, schemaKeywords.email)) {
    return zodKeywordMapper.email(shouldCoerce(options.coercion, 'strings'))
  }

  if (isKeyword(current, schemaKeywords.url)) {
    return zodKeywordMapper.url(shouldCoerce(options.coercion, 'strings'))
  }

  if (isKeyword(current, schemaKeywords.number)) {
    return zodKeywordMapper.number(shouldCoerce(options.coercion, 'numbers'))
  }

  if (isKeyword(current, schemaKeywords.integer)) {
    return zodKeywordMapper.integer(shouldCoerce(options.coercion, 'numbers'))
  }

  if (isKeyword(current, schemaKeywords.min)) {
    return zodKeywordMapper.min(current.args)
  }
  if (isKeyword(current, schemaKeywords.max)) {
    return zodKeywordMapper.max(current.args)
  }

  if (isKeyword(current, schemaKeywords.datetime)) {
    return zodKeywordMapper.datetime(current.args.offset, current.args.local)
  }

  if (isKeyword(current, schemaKeywords.date)) {
    return zodKeywordMapper.date(current.args.type, shouldCoerce(options.coercion, 'dates'))
  }

  if (isKeyword(current, schemaKeywords.time)) {
    return zodKeywordMapper.time(current.args.type, shouldCoerce(options.coercion, 'dates'))
  }

  if (current.keyword in zodKeywordMapper && 'args' in current) {
    const value = zodKeywordMapper[current.keyword as keyof typeof zodKeywordMapper] as (typeof zodKeywordMapper)['const']

    return value((current as SchemaKeywordBase<unknown>).args as any)
  }

  if (isKeyword(current, schemaKeywords.optional)) {
    if (siblings.some((schema) => isKeyword(schema, schemaKeywords.default))) return ''

    return value()
  }

  if (current.keyword in zodKeywordMapper) {
    return value()
  }

  return undefined
}
