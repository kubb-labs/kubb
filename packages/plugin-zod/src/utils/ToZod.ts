/**
 * See https://github.com/colinhacks/tozod/blob/master/src/index.ts
 * Adapted based on https://github.com/colinhacks/zod/issues/372
 */

import type { z } from 'zod'

type isAny<T> = [any extends T ? 'true' : 'false'] extends ['true'] ? true : false
type nonoptional<T> = T extends undefined ? never : T
type nonnullable<T> = T extends null ? never : T
type equals<X, Y> = [X] extends [Y] ? ([Y] extends [X] ? true : false) : false

type zodKey<T> = isAny<T> extends true
  ? 'any'
  : equals<T, boolean> extends true //[T] extends [booleanUtil.Type]
    ? 'boolean'
    : [undefined] extends [T]
      ? 'optional'
      : [null] extends [T]
        ? 'nullable'
        : T extends any[]
          ? 'array'
          : equals<T, string> extends true
            ? 'string'
            : equals<T, bigint> extends true //[T] extends [bigintUtil.Type]
              ? 'bigint'
              : equals<T, number> extends true //[T] extends [numberUtil.Type]
                ? 'number'
                : equals<T, Date> extends true //[T] extends [dateUtil.Type]
                  ? 'date'
                  : T extends { [k: string]: any } //[T] extends [structUtil.Type]
                    ? 'object'
                    : 'rest'

export type ToZod<T> = {
  any: z.ZodAny
  optional: z.ZodOptional<ToZod<nonoptional<T>>>
  nullable: z.ZodNullable<ToZod<nonnullable<T>>>
  array: T extends Array<infer U> ? z.ZodArray<ToZod<U>> : never
  string: z.ZodString
  bigint: z.ZodBigInt
  number: z.ZodNumber
  boolean: z.ZodBoolean
  date: z.ZodDate

  object: z.ZodObject<
    // @ts-expect-error cannot convert without Extract but Extract removes the type
    {
      [K in keyof T]: T[K]
    },
    'passthrough',
    z.ZodTypeAny,
    T
  >
  rest: z.ZodType<T>
}[zodKey<T>]

export type ZodShape<T> = {
  // Require all the keys from T
  [key in keyof T]-?: ToZod<T[key]>
}
