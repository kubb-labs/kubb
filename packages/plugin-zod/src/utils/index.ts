export type { ToZod, ZodShape } from './ToZod.ts'
