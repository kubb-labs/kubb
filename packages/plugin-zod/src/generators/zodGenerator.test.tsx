import { matchFiles, mockedPluginManager } from '@kubb/core/mocks'

import path from 'node:path'
import type { Plugin } from '@kubb/core'
import type { HttpMethod } from '@kubb/oas'
import { parse } from '@kubb/oas/parser'
import { OperationGenerator, SchemaGenerator } from '@kubb/plugin-oas'
import { getSchemas } from '@kubb/plugin-oas/utils'
import type { PluginZod } from '../types.ts'
import { zodGenerator } from './zodGenerator.tsx'

describe('zodGenerator schema', async () => {
  const testData = [
    {
      name: 'Pet',
      input: '../../mocks/petStore.yaml',
      options: {},
    },
    {
      name: 'Pets',
      input: '../../mocks/petStore.yaml',
      options: {
        typedSchema: true,
      },
    },
    {
      name: 'Pet',
      input: '../../mocks/petStore.yaml',
      options: {},
    },
    {
      name: 'Pet',
      input: '../../mocks/petStore.yaml',
      options: {
        coercion: true,
      },
    },
    {
      name: 'PetTupleObject',
      input: '../../mocks/petStore.yaml',

      options: {},
    },
    {
      name: 'OptionalPet',
      input: '../../mocks/petStore.yaml',
      options: {
        typed: true,
      },
    },
    {
      name: 'OptionalPet',
      input: '../../mocks/petStore.yaml',
      options: {
        typed: true,
      },
    },
    {
      name: 'PetArray',
      input: '../../mocks/petStore.yaml',
      options: {},
    },
    {
      name: 'Order',
      input: '../../mocks/petStore.yaml',
      options: {},
    },
    {
      name: 'Order',
      input: '../../mocks/petStore.yaml',
      options: {
        dateType: 'string',
      },
    },
    {
      name: 'Order',
      input: '../../mocks/petStore.yaml',
      options: {
        dateType: false,
      },
    },
    {
      name: 'UuidSchema',
      input: '../../mocks/constCases.yaml',
      options: {},
    },
    {
      name: 'NullableString',
      input: '../../mocks/constCases.yaml',
      options: {},
    },
    {
      name: 'NullableStringWithAnyOf',
      input: '../../mocks/constCases.yaml',
      options: {},
    },
    {
      name: 'NullableStringUuid',
      input: '../../mocks/constCases.yaml',
      options: {},
    },
    {
      name: 'StringValueConst',
      input: '../../mocks/constCases.yaml',
      options: {},
    },
    {
      name: 'NumberValueConst',
      input: '../../mocks/constCases.yaml',
      options: {},
    },
    {
      name: 'MixedValueTypeConst',
      input: '../../mocks/constCases.yaml',
      options: {},
    },
    {
      name: 'enumVarNames.Type',
      input: '../../mocks/enums.yaml',

      options: {},
    },
    {
      name: 'enumNames.Type',
      input: '../../mocks/enums.yaml',
      options: {},
    },
    {
      name: 'enumNullable',
      input: '../../mocks/enums3_1.yaml',
      options: {},
    },
    {
      name: 'Example',
      input: '../../mocks/recursive.yaml',
      options: {},
    },
    {
      name: 'test',
      input: '../../mocks/anyof.yaml',
      options: {},
    },
    {
      name: 'test',
      input: '../../mocks/oneof.yaml',
      options: {},
    },
    {
      name: 'Example',
      input: '../../mocks/lazy.yaml',
      options: {},
    },
  ] as const satisfies Array<{
    input: string
    name: string
    options: Partial<PluginZod['resolvedOptions']>
  }>

  test.each(testData)('$name', async (props) => {
    const oas = await parse(path.resolve(__dirname, props.input))

    const options: PluginZod['resolvedOptions'] = {
      dateType: 'date',
      transformers: {},
      typed: false,
      typedSchema: false,
      unknownType: 'any',
      mapper: {},
      importPath: 'zod',
      coercion: false,
      operations: false,
      ...props.options,
    }
    const plugin = { options } as Plugin<PluginZod>
    const instance = new SchemaGenerator(options, {
      oas,
      pluginManager: mockedPluginManager,
      plugin,
      contentType: 'application/json',
      include: undefined,
      override: undefined,
      mode: 'split',
      output: './gen',
    })
    await instance.build(zodGenerator)

    const schemas = getSchemas({ oas })
    const name = props.name
    const schema = schemas[name]!
    const tree = instance.parse({ schema, name })

    const files = await zodGenerator.schema?.({
      schema,
      name,
      tree,
      options,
      instance,
    })

    await matchFiles(files)
  })
})

describe('zodGenerator operation', async () => {
  const testData = [
    {
      name: 'showPetById',
      input: '../../mocks/petStore.yaml',
      path: '/pets/{petId}',
      method: 'get',
      options: {},
    },
    {
      name: 'getPets',
      input: '../../mocks/petStore.yaml',
      path: '/pets',
      method: 'get',
      options: {},
    },
    {
      name: 'createPet',
      input: '../../mocks/petStore.yaml',
      path: '/pets',
      method: 'post',
      options: {},
    },
    {
      name: 'createPet with unknownType any',
      input: '../../mocks/petStore.yaml',
      path: '/pets',
      method: 'post',
      options: {
        unknownType: 'any',
      },
    },
    {
      name: 'deletePet',
      input: '../../mocks/petStore.yaml',
      path: '/pets/{petId}',
      method: 'delete',
      options: {},
    },
  ] as const satisfies Array<{
    input: string
    name: string
    path: string
    method: HttpMethod
    options: Partial<PluginZod['resolvedOptions']>
  }>

  test.each(testData)('$name', async (props) => {
    const oas = await parse(path.resolve(__dirname, props.input))

    const options: PluginZod['resolvedOptions'] = {
      dateType: 'date',
      transformers: {},
      typed: false,
      typedSchema: false,
      unknownType: 'any',
      mapper: {},
      importPath: 'zod',
      coercion: false,
      operations: false,
      ...props.options,
    }
    const plugin = { options } as Plugin<PluginZod>
    const instance = new OperationGenerator(options, {
      oas,
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin,
      contentType: undefined,
      override: undefined,
      mode: 'split',
      exclude: [],
    })
    const operation = oas.operation(props.path, props.method)
    const files = await zodGenerator.operation?.({
      operation,
      options,
      instance,
    })

    await matchFiles(files)
  })
})
