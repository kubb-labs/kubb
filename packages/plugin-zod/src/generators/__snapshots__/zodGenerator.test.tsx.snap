// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zodGenerator operation > 'createPet with unknownType any' 1`] = `
"/**
 * @description Null response
 */
export const createPets201 = z.any()

/**
 * @description unexpected error
 */
export const createPetsError = z.lazy(() => error)

export const createPetsMutationRequest = z.object({"name": z.string(),"tag": z.string()})

export const createPetsMutationResponse = z.any()"
`;

exports[`zodGenerator operation > 'createPet' 1`] = `
"/**
 * @description Null response
 */
export const createPets201 = z.any()

/**
 * @description unexpected error
 */
export const createPetsError = z.lazy(() => error)

export const createPetsMutationRequest = z.object({"name": z.string(),"tag": z.string()})

export const createPetsMutationResponse = z.any()"
`;

exports[`zodGenerator operation > 'deletePet' 1`] = `"export const deletePetsPetidMutationResponse = z.any()"`;

exports[`zodGenerator operation > 'getPets' 1`] = `
"export const listPetsQueryParams = z.object({"limit": z.string().describe("How many items to return at one time (max 100)").optional()}).optional()

/**
 * @description A paged array of pets
 */
export const listPets200 = z.lazy(() => pets)

/**
 * @description unexpected error
 */
export const listPetsError = z.lazy(() => error)

/**
 * @description A paged array of pets
 */
export const listPetsQueryResponse = z.lazy(() => pets)"
`;

exports[`zodGenerator operation > 'showPetById' 1`] = `
"export const showPetByIdPathParams = z.object({"petId": z.string().describe("The id of the pet to retrieve"),"testId": z.string().describe("The id of the pet to retrieve")})

/**
 * @description Expected response to a valid request
 */
export const showPetById200 = z.lazy(() => pet)

/**
 * @description unexpected error
 */
export const showPetByIdError = z.lazy(() => error)

/**
 * @description Expected response to a valid request
 */
export const showPetByIdQueryResponse = z.lazy(() => pet)"
`;

exports[`zodGenerator schema > 'Example' 1`] = `"export const example = z.object({"name": z.string(),"children": z.array(z.lazy(() => example))})"`;

exports[`zodGenerator schema > 'Example' 2`] = `"export const example = z.object({"nestedExamples": z.lazy(() => example).optional()})"`;

exports[`zodGenerator schema > 'MixedValueTypeConst' 1`] = `
"/**
 * @description This probably should fail miserably
 */
export const mixedValueTypeConst = z.object({"foobar": z.literal("foobar")}).describe("This probably should fail miserably")"
`;

exports[`zodGenerator schema > 'NullableString' 1`] = `"export const nullableString = z.string().nullable()"`;

exports[`zodGenerator schema > 'NullableStringUuid' 1`] = `"export const nullableStringUuid = z.string().uuid().nullable()"`;

exports[`zodGenerator schema > 'NullableStringWithAnyOf' 1`] = `"export const nullableStringWithAnyOf = z.union([z.string(), z.null()])"`;

exports[`zodGenerator schema > 'NumberValueConst' 1`] = `
"/**
 * @description its value is equal to the value of the keyword
 */
export const numberValueConst = z.object({"foobar": z.literal(42)}).describe("its value is equal to the value of the keyword")"
`;

exports[`zodGenerator schema > 'OptionalPet' 1`] = `"export const optionalPet = z.object({"id": z.number().int().optional(),"name": z.string().optional(),"tag": z.string().optional()}) as z.ZodType<OptionalPet>"`;

exports[`zodGenerator schema > 'OptionalPet' 2`] = `"export const optionalPet = z.object({"id": z.number().int().optional(),"name": z.string().optional(),"tag": z.string().optional()}) as z.ZodType<OptionalPet>"`;

exports[`zodGenerator schema > 'Order' 1`] = `"export const order = z.object({"id": z.number().int().optional(),"petId": z.number().int().optional(),"quantity": z.number().int().optional(),"shipDate": z.date().optional(),"status": z.enum(["placed", "approved", "delivered"]).describe("Order Status").optional(),"complete": z.boolean().optional()})"`;

exports[`zodGenerator schema > 'Order' 2`] = `"export const order = z.object({"id": z.number().int().optional(),"petId": z.number().int().optional(),"quantity": z.number().int().optional(),"shipDate": z.string().datetime().optional(),"status": z.enum(["placed", "approved", "delivered"]).describe("Order Status").optional(),"complete": z.boolean().optional()})"`;

exports[`zodGenerator schema > 'Order' 3`] = `"export const order = z.object({"id": z.number().int().optional(),"petId": z.number().int().optional(),"quantity": z.number().int().optional(),"shipDate": z.string().optional(),"status": z.enum(["placed", "approved", "delivered"]).describe("Order Status").optional(),"complete": z.boolean().optional()})"`;

exports[`zodGenerator schema > 'Pet' 1`] = `"export const pet = z.object({"id": z.number().int(),"name": z.string(),"date": z.date().optional(),"tag": z.string().min(5).max(100).optional()})"`;

exports[`zodGenerator schema > 'Pet' 2`] = `"export const pet = z.object({"id": z.number().int(),"name": z.string(),"date": z.date().optional(),"tag": z.string().min(5).max(100).optional()})"`;

exports[`zodGenerator schema > 'Pet' 3`] = `"export const pet = z.object({"id": z.coerce.number().int(),"name": z.coerce.string(),"date": z.coerce.date().optional(),"tag": z.coerce.string().min(5).max(100).optional()})"`;

exports[`zodGenerator schema > 'PetArray' 1`] = `
"/**
 * @description List of Pet object
 */
export const petArray = z.array(z.lazy(() => pet)).min(1).max(3).describe("List of Pet object")"
`;

exports[`zodGenerator schema > 'PetTupleObject' 1`] = `
"/**
 * @description Tuple of exact length 2 nested in an object
 */
export const petTupleObject = z.object({"tupleProperty": z.tuple([z.string(), z.string()]).optional()}).describe("Tuple of exact length 2 nested in an object")"
`;

exports[`zodGenerator schema > 'Pets' 1`] = `
"export const pets = z.array(z.object({"id": z.number().int(),"name": z.string(),"tag": z.string().optional()}))

export type Pets = z.infer<typeof pets>"
`;

exports[`zodGenerator schema > 'StringValueConst' 1`] = `
"/**
 * @description its value is equal to the value of the keyword
 */
export const stringValueConst = z.object({"foobar": z.literal("foobar")}).describe("its value is equal to the value of the keyword")"
`;

exports[`zodGenerator schema > 'UuidSchema' 1`] = `"export const uuidSchema = z.string().uuid()"`;

exports[`zodGenerator schema > 'enumNames.Type' 1`] = `"export const enumNamesType = z.enum(["0", "1"])"`;

exports[`zodGenerator schema > 'enumNullable' 1`] = `"export const enumNullable = z.enum(["Pending", "Received"]).nullable()"`;

exports[`zodGenerator schema > 'enumVarNames.Type' 1`] = `"export const enumVarNamesType = z.union([z.literal(0), z.literal(1)])"`;

exports[`zodGenerator schema > 'test' 1`] = `"export const test = z.union([z.object({"propertyA": z.string()}).strict(), z.object({"propertyA": z.string(),"propertyB": z.string()}).strict()])"`;

exports[`zodGenerator schema > 'test' 2`] = `"export const test = z.union([z.object({"propertyA": z.string().optional()}), z.object({"propertyA": z.string().optional()})])"`;
